js-api-spec/node-package-importer.node.test.ts

```diff
@@ -0,0 +1,580 @@
+        input: '@use "pkg:foo";',
+        output: 'a {b: c;}',
+        files: {
+          'node_modules/foo/src/sass/_styles.scss': 'a {b: c}',
+          'node_modules/foo/package.json': JSON.stringify({
+            sass: 'src/sass/_styles.scss',
+          }),
+        },
+      }));
+
+    it('style key in package.json', () =>
+      testPackageImporter({
+        input: '@use "pkg:foo";',
+        output: 'a {b: c;}',
+        files: {
+          'node_modules/foo/src/sass/_styles.scss': 'a {b: c}',
+          'node_modules/foo/package.json': JSON.stringify({
+            style: 'src/sass/_styles.scss',
+          }),
+        },
+      }));
+
+    ['index.scss', 'index.css', '_index.scss', '_index.css'].forEach(
+      fileName => {
+        it(`loads from ${fileName}`, () =>
+          testPackageImporter({
+            input: '@use "pkg:foo";',
+            output: 'a {b: c;}',
+            files: {
+              [`node_modules/foo/${fileName}`]: 'a {b: c}',
+              'node_modules/foo/package.json': JSON.stringify({}),
+            },
+          }));
+      }
+    );
+    ['index.sass', '_index.sass'].forEach(fileName => {
+      it(`loads from ${fileName}`, () =>
+        testPackageImporter({
+          input: '@use "pkg:foo";',
+          output: 'a {b: c;}',
+          files: {
+            [`node_modules/foo/${fileName}`]: 'a \n b: c',
+            'node_modules/foo/package.json': JSON.stringify({}),
+          },
+        }));
+    });
+  });
+
+  it('with subpath, resolves relative to package root', () =>
+    testPackageImporter({
+      input: '@use "pkg:bar/src/styles/sass";',
+      output: 'a {b: c;}',
+      files: {
+        'node_modules/bar/src/styles/sass/index.scss': 'a {b: c}',
+        'node_modules/bar/package.json': JSON.stringify({}),
+      },
+    }));
+
+  describe('resolves from packages', () => {
+    it('resolves from secondary @use', () =>
+      sandbox(dir => {
+        dir.write({
+          'node_modules/bah/index.scss': 'a {b: c}',
+          'node_modules/bah/package.json': JSON.stringify({}),
+          '_vendor.scss': '@use "pkg:bah";',
+        });
+        dir.chdir(() => {
+          const result = compileString('@use "vendor";', {
+            importers: [
+              nodePackageImporter,
+              {
+                findFileUrl: file => dir.url(file),
+              },
+            ],
+          });
+          expect(result.css).toEqualIgnoringWhitespace('a {b: c;}');
+        });
+      }));
+
+    it('resolves from secondary @use pkg root', () =>
+      testPackageImporter({
+        input: '@use "pkg:bah";',
+        output: 'a {b: c;}',
+        files: {
+          'node_modules/bah/index.scss': '@use "pkg:bar";',
+          'node_modules/bah/package.json': JSON.stringify({}),
+          'node_modules/bar/index.scss': 'a {b: c}',
+          'node_modules/bar/package.json': JSON.stringify({}),
+        },
+      }));
+
+    it('relative import in package', () =>
```

Comment:
This is a failing test that reflects the setup that exists in Bootstrap, and exposes an issue that's puzzling me a bit. 

It is able to find the package `foo`, but from `node_modules/foo/scss/foo.scss`, `@use "mixins/banner";` is unable to find the path `node_modules/foo/scss/mixins/_banner.scss`. 

I tested to confirm this isn't due to partials or extension matching. The Node Package Importer seems to be impacting the subsequent resolution of relative paths. I suspect there's a breaking issue in dart-sass around the ImporterCache.

Reply:
That definitely seems like a bug, glad you caught it in a test. Here's what _should_ be happening:

* `pkg:foo` gets passed to `NodePackageImporter.canonicalize()`, which returns `file://.../node_modules/foo/scss/foo.scss`.
* `foo.scss` gets loaded with that canonical URL and `NodePackageImporter` as the current importer to handle relative URLs.
* The import cache resolves `mixins/banner` relative to that canonical URL to return `file://.../node_modules/foo/scss/mixins/banner`.
* That resolved URL then gets passed to `NodePackageImporter.canonicalize()`, which sees that it's a `file:` URL and passes it to `FilesystemImporter.canonicalize()`.
* `FilesystemImporter.canonicalize()` returns `file://.../node_modules/foo/scss/mixins/_banner.scss`.
* `_banner.scss` is loaded and everything is great.